### PR TITLE
Support Interrupt Remapping and Queued Invalidation in IOMMU

### PR DESCRIPTION
**Pull request by sdww0**
Replace #961

### CODE REVIEW

**Review by tatetian — COMMENTED** [review]
Good job on implementing interrupt remapping! It is important to guard against (potential) malicious interrupts from peripheral devices. IOMMU is not enabled in CI until a recent PR #1537 . Could you double check that this IOMMU feature is indeed enabled in the CI test?

File: ostd/src/bus/pci/capability/msix.rs [hunk 1]
```diff
@@ -163,18 +164,41 @@ impl CapabilityMsixData {
...
-            .unwrap();
-        let old_handles = core::mem::replace(&mut self.irqs[index as usize], Some(handle));
+        if has_interrupt_remapping() {
+            let mut handle = irq.inner_irq().bind_remapping_entry().unwrap().lock();
+            // Enable irt entry
+            let irt_entry_mut = handle.irt_entry_mut().unwrap();
+            irt_entry_mut.enable_default(irq.num() as u32);
+
+            // Use remappable format
+            let mut address = 0xFEE0_0000u32;
```
[the remark this answers is not in the document]
Comment (sdww0) [reply ostd/src/bus/pci/capability/msix.rs]: This is MSIX's default address for x86 architecture. I have stored this value in a constant variable `MSIX_DEFAULT_MSG_ADDR`.

File: ostd/src/bus/pci/capability/msix.rs [hunk 2]
```diff
@@ -163,18 +164,41 @@ impl CapabilityMsixData {
...
+            address |= 0b1_1000;
+            address |= (handle.index() as u32 & 0x7FFF) << 5;
+            address |= (handle.index() as u32 & 0x8000) >> 13;
```
[the remark this answers is not in the document]
Comment (sdww0) [reply ostd/src/bus/pci/capability/msix.rs:188]: Of course, In the newest commit, I review these places and add comments to explain the reason or the name of these values.

I hope the documentation is better understood than the previous one.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**